library/src/org/wordpress/passcodelock/DefaultAppLock.java

```diff
@@ -12,125 +11,183 @@
+import android.os.Build;
 import android.os.Bundle;
 import android.preference.PreferenceManager;
+import android.text.TextUtils;
 import android.util.Base64;
 
-import org.wordpress.passcodelock.BuildConfig;
-
 public class DefaultAppLock extends AbstractAppLock {
-
-    private Application currentApp; //Keep a reference to the app that invoked the locker
-    private SharedPreferences settings;
-    private Date lostFocusDate;
-
-    //Add back-compatibility
+    private static final String UNLOCK_CLASS_NAME = PasscodeUnlockActivity.class.getName();
     private static final String OLD_PASSWORD_SALT = "sadasauidhsuyeuihdahdiauhs";
     private static final String OLD_APP_LOCK_PASSWORD_PREF_KEY = "wp_app_lock_password_key";
 
-    public DefaultAppLock(Application currentApp) {
+    private Application mCurrentApp;
+    private SharedPreferences mSharedPreferences;
+    private Date mLostFocusDate;
+
+    public DefaultAppLock(Application app) {
         super();
-        SharedPreferences settings = PreferenceManager.getDefaultSharedPreferences(currentApp);
-        this.settings = settings;
-        this.currentApp = currentApp;
+        mCurrentApp = app;
+        mSharedPreferences = PreferenceManager.getDefaultSharedPreferences(mCurrentApp);
     }
 
-    public void enable(){
-    	if (android.os.Build.VERSION.SDK_INT < 14)
-    		return;
+    /** {@link PasscodeUnlockActivity} is always exempt. */
+    @Override
+    public boolean isExemptActivity(String activityName) {
+        return !UNLOCK_CLASS_NAME.equals(activityName) && super.isExemptActivity(activityName);
+    }
 
-        if( isPasswordLocked() ) {
-            currentApp.unregisterActivityLifecycleCallbacks(this);
-            currentApp.registerActivityLifecycleCallbacks(this);
+    @Override
+    public void onActivityPaused(Activity activity) {
+        if (!isExemptActivity(activity.getClass().getName())) mLostFocusDate = new Date();
+    }
+
+    @Override
+    public void onActivityResumed(Activity activity) {
+        if (!isExemptActivity(activity.getClass().getName()) && shouldShowUnlockScreen()) {
+            Intent i = new Intent(activity.getApplicationContext(), PasscodeUnlockActivity.class);
+            i.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+            activity.getApplication().startActivity(i);
         }
     }
 
-    public void disable( ){
-    	if (android.os.Build.VERSION.SDK_INT < 14)
-    		return;
+    @Override public void onActivityCreated(Activity arg0, Bundle arg1) {}
+    @Override public void onActivityDestroyed(Activity arg0) {}
+    @Override public void onActivitySaveInstanceState(Activity arg0, Bundle arg1) {}
+    @Override public void onActivityStarted(Activity arg0) {}
+    @Override public void onActivityStopped(Activity arg0) {}
 
-        currentApp.unregisterActivityLifecycleCallbacks(this);
+    public void enable() {
+    	if (Build.VERSION.SDK_INT < Build.VERSION_CODES.ICE_CREAM_SANDWICH) return;
```

Comment:
Can we re-use `isSupportedApi()` of `AppLockManager` in some way? Maybe move that method in this class....